### ISSUE TITLE
Serial build is slow: would there be ways to speed it up?

### ISSUE THREAD

**Issue by aurelien-reeves** [cucumber/common#1646]
The serial build is slow. It may last for about 50 minutes.

I don't like to merge PR before all jobs are done. Sometimes it is difficult because of changes on the main branches which need to be merged in the PR, then waiting for the serial build to finish and hoping no other changes would be done on main in the meantime...

I know, I have the possibility to force merge thanks to my admin authorization. But this remain a workaround. If we could prevent it, that would be great.

So I suggest to try speeding-up the CI. Any ideas?

My first idea would be to split that serial build into languages-specific builds. That way we would have parallel builds per modules, and parallel builds per languages.

Would we have specific requirements that would not be met with such parallelization?

Would you have other ideas?

**Comment by mpkorstanje** [comment]
There is an implicit build order encoded in the make file. We re-encode that order in our parallel build through dependencies. When the make file order is wrong, we can't build locally and this has broken often in the past.

If we de remove the serial build, we do need something to confirm that building locally will still work.

**Comment by aurelien-reeves** [comment]
I understand the need for the serial build. I suggest to split it per languages: one build for all java modules, one for javascript, ...

I think that it should cover the need you described?

**Comment by davidjgoss** [comment]
I think there's some cross-language dependency in places - e.g. the java and ruby versions of `html-formatter` need the javascript one to have built.

**Comment by aurelien-reeves** [comment]
Thanks for the info

**Comment by mattwynne** [comment]
https://github.com/cucumber/build/issues/18 is somewhat related to this.

**Comment by aurelien-reeves** [comment]
> 
> 
> [cucumber/build#18](https://github.com/cucumber/build/issues/18) is somewhat related to this.
> 
> @ciaranmcnulty and I have had one pairing session on it, so it's too early to say what impact it could have.

I am not sure that would be related.
I think the serial build is slow because we build everything in every language without any parallelization.

**Comment by aslakhellesoy** [comment]
I came over this today - something to consider?

https://earthly.dev/

**Comment by mpkorstanje** [comment]
Looks very interesting, potentially a general improvement.

Its parallism doesn't appear to be distributable across multiple nodes. So maybe not a solution to ensure the local build keeps working. 

However if the dependency hierarchy is simple enough, we may not have to.

**Comment by ehuelsmann** [comment]
The Perl step is one of the steps that take longest to run. I've submitted changes to the Dockerfile which could reduce Perl build time. I'm not sure where *inside* the Perl job the time is spent exactly (and why that would take this amount of time when the e.g. Ruby one does not). However, no amount of reducing Perl time will significantly reduce the 50 minutes. There are just a lot of components here...

**Comment by mattwynne** [comment]
What about caching `make` artefacts using https://github.com/actions/cache? Could that work? Maybe we could do `make clean` in a nightly build or something?

**Comment by aurelien-reeves** [comment]
At the moment the CI is using CircleCI, not github actions.
But they certainly have some cache mechanism too 🤔

**Comment by mpkorstanje** [comment]
That is putting a lot of trust in the correctness of our make files and once broken, almost impossible to debug.

**Comment by aurelien-reeves** [comment]
Yes. I must say that I like the serial build being totally fresh each times. Using a cache, and it is already not idempotent anymore

**Comment by mpkorstanje** [comment]
Yes.
